torch/distributed/elastic/timer/file_based_local_timer.py

```diff
@@ -209,117 +204,97 @@ def run_once(self) -> None:
-                self._log_event("kill worker process", expired_timer)
+            # clear metadata for reaped worker to avoid keep reaping in future
+            self._clients_metadata[etimer.scope_id] = None
+            if self._reap_worker(etimer.worker_pid, etimer.signal):
+                log.info("Successfully reaped worker=[%s] with signal=%s", etimer.worker_pid, etimer.signal)
+                self._log_event("kill worker process", etimer)
             else:
-                log.error("Error reaping worker=[%s]. Will retry on next watchdog.", worker_pid)
-        self.clear_timers(reaped_worker_pids)
-
-    def _get_scopes(self, timer_requests: List[FileTimerRequest]) -> List[str]:
-        return [r.scope_id for r in timer_requests]
+                log.error("Error reaping worker=[%s]. Will retry on next watchdog.", etimer.worker_pid)
+        self.clear_timers(reaped_workers)
 
-    def _get_requests(self, fd: io.TextIOWrapper, max_interval: float) -> List[FileTimerRequest]:
+    def _get_requests(self, max_interval: float) -> List[FileTimerRequest]:
         start = time.time()
         requests = []
         while not self._stop_signaled or self._run_once:
-            # For named pipe, readline() is blocking when at least one writer opens.
-            # It returns only when flush() is called at the writer side.
-            # Note that flush() is automatically called inside close().
-            # After the last writer closes, readline() is not blocking.
-            # It will return an empty string when it's at end-of-file.
-            # Since the client side always opens the pipe, writes a message and closes
-            # the pipe immediately, the readline() call below is not blocking for long.
-            json_request = fd.readline()
-            if len(json_request) == 0:
-                if self._run_once:
-                    break
-                time.sleep(min(max_interval, 1))
-            else:
-                request = json.loads(json_request)
-                pid = request["pid"]
-                scope_id = request["scope_id"]
-                expiration_time = request["expiration_time"]
-                signal = request["signal"]
+            for metadata in self._clients_metadata.values():
+                if not metadata:
+                    continue
+                last_mtime = os.path.getmtime(f"{self._dir_path}/{metadata.scope_id}")
                 requests.append(
                     FileTimerRequest(
-                        worker_pid=pid, scope_id=scope_id, expiration_time=expiration_time, signal=signal
+                        worker_pid=metadata.worker_pid,
+                        scope_id=metadata.scope_id,
+                        expiration_time=last_mtime + metadata.expiration_time,
+                        signal=metadata.signal,
                     )
                 )
+            time.sleep(min(max_interval, 10))
```

Comment:
should we be using inotify instead of sleeping?

Reply:
I can explore inotify, seems more cleaner approach